pkg/controller/usersignup/approval.go

```diff
@@ -0,0 +1,120 @@
+		return false, "", nil
+	}
+
+	status := &toolchainv1alpha1.ToolchainStatus{}
+	if err := cl.Get(context.TODO(), types.NamespacedName{Namespace: userSignup.Namespace, Name: crtConfig.GetToolchainStatusName()}, status); err != nil {
+		return false, "", errors.Wrapf(err, "unable to read ToolchainStatus resource")
+	}
+	counts, err := counter.GetCounts()
+	if err != nil {
+		return false, "", errors.Wrapf(err, "unable to get the number of provisioned users")
+	}
+
+	targetCluster := getOptimalTargetCluster(userSignup, getMemberClusters, hasNotReachedMaxNumberOfUsersThreshold(config, counts), hasEnoughResources(config, status))
+	if targetCluster == "" {
+		return false, "", fmt.Errorf("no suitable member cluster found")
+	}
+
+	// approved manually
+	if userSignup.Spec.Approved {
+		return true, targetCluster, nil
+	}
+
+	if config.Spec.AutomaticApproval.MaxNumberOfUsers.Overall != 0 {
+
+		if config.Spec.AutomaticApproval.MaxNumberOfUsers.Overall <= counts.MasterUserRecordCount {
+			return false, "", nil
+		}
+	}
+
+	if config.Spec.AutomaticApproval.ResourceCapacityThreshold.DefaultThreshold != 0 {
+		if config.Spec.AutomaticApproval.MaxNumberOfUsers.Overall <= counts.MasterUserRecordCount {
+			return false, "", nil
+		}
+	}
+	return true, targetCluster, nil
+}
+
+func hasNotReachedMaxNumberOfUsersThreshold(config *toolchainv1alpha1.HostOperatorConfig, counts counter.Counts) cluster.Condition {
+	return func(cluster *cluster.CachedToolchainCluster) bool {
+		numberOfUserAccounts := counts.UserAccountsPerClusterCounts[cluster.Name]
+		threshold := config.Spec.AutomaticApproval.MaxNumberOfUsers.SpecificPerMemberCluster[cluster.Name]
+		return threshold == 0 || numberOfUserAccounts <= threshold
+	}
+}
+
+func hasEnoughResources(config *toolchainv1alpha1.HostOperatorConfig, status *toolchainv1alpha1.ToolchainStatus) cluster.Condition {
+	return func(cluster *cluster.CachedToolchainCluster) bool {
+		threshold, found := config.Spec.AutomaticApproval.ResourceCapacityThreshold.SpecificPerMemberCluster[cluster.Name]
+		if !found {
+			threshold = config.Spec.AutomaticApproval.ResourceCapacityThreshold.DefaultThreshold
+		}
+		if threshold == 0 {
+			return true
+		}
+		for _, memberStatus := range status.Status.Members {
+			if memberStatus.ClusterName == cluster.Name {
+				if len(memberStatus.MemberStatus.ResourceUsage.MemoryUsagePerNodeRole) > 0 {
+					for _, usagePerNode := range memberStatus.MemberStatus.ResourceUsage.MemoryUsagePerNodeRole {
+						if usagePerNode >= threshold {
+							return false
+						}
+					}
+					return true
+				}
+				return false
+			}
+		}
+		return false
+	}
+}
+
+func getOptimalTargetCluster(userSignup *toolchainv1alpha1.UserSignup, getMemberClusters cluster.GetMemberClustersFunc, conditions ...cluster.Condition) string {
+	// If a target cluster hasn't been selected, select one from the members
+	if userSignup.Spec.TargetCluster != "" {
+		return userSignup.Spec.TargetCluster
+	} else {
```

Comment:
this else clause can probably be removed?

Reply:
good catch 7c8e20e